CONFIG.md
# CONFIG FILES
Configuration files are used for specifying model and hyper parameters for the run, each parameter follows the format:

```
# Comment, will not be parsed
parameter_name = <value>
``` 

Values can be of different types. Below are a list of parameters, their type and a short description. For all parameters default value is also specified.

## RUN PARAMS

| Parameter | type | Description | Default Value |
| --- | --- | --- | --- |
| name | string | run name used for experimentID| "" |
| hide_ui | boolean | If true, ProteinCNN will not start webserver for live data plotting | False |
| min_update | int | minimum number of minibatches processed before training can terminate  | 1000 |
| max_update | int | training terminate if max_update  == minibatches processed | 2000 |
| minibatch_size | int | ... | 1 |
| eval_interval | int | number of training steps before model is evaluated again on validation dataset | 10 |
| learning_rate | float | model learning rate at each training step | 0.01 |
| max_sequence_length | int | max protein sequence length used in for run | 2000 |
| use_evolutionary | boolean | if true, the input will consist of the amino-acids and their PSSM values | True |
| training_file | string | raw data file used for training samples | '' |
| validation_file | string | raw data file used for evaluation model | '' |
| max_time | int | maximum time (in seconds). If sat the training will terminate if the time used reaches it | None |
| angles_lr | float | specify unique learning rate for the learned angle matrix | 0.01 |
| loss_atoms | string | either 'all' or 'c_alpha'. Determines if drmsd loss is calculated on the c_alpha backbone molecyle or all molecyles in the backbone. | 'c_alpha' |
| weight_decay |float | specify the weight decay of the optimizer (AdamOptimizer) | 0.0 |
| betas | float,float | specify betas for the optimizer | [0.95,0.99] |
| print_model_summary | boolean | prints are complete model summary of the network (does not work for RNN) | False |
| milestones | [(int,int),...] | specify validation milestones for the run. Each entry is given as a tuple '(100,12.5),(500,11.3)' | None |
| restart_on_failed_milestone | boolean | restarts model training if milestone is not reached, same parameters but using new seed | False |
| max_restarts | int | maximum number of run restarts | 0 |

## MODEL PARAMS
| Parameter | type | Description | Default Value |
| --- | --- | --- | --- |
| architecture | string | specify which architecture is used; either [cnn,rnn,resnet] | cnn |
| dropout | [int,int] | specifies input dropout and dropout between layers | [0.5,0.5] |
| output_size | int | output_size for the model, a alphabet of same size will be created for angle classification | 60 |


## CNN PARAMS
| Parameter | type | Description | Default Value |
| --- | --- | --- | --- |
| kernel | [int,...?] | kernel for each convolution layer, if single value is given it will be used for all layers | 11 |
| spatial_dropout | boolean | specify if spatial_dropout is used instead of normal dropout | False |
| layers | int | number of layers | 4 |
| dilation | [int,...?] | dilation for each convolution layer, if single value is given it will be used for all layers | 1 |
| stride | [int,...?]| stride for each convolution layer, if single value is given it will be used for all layers | 1 |
| channels | [int,...?] | channels for each convolution layer, if single value is given it will be used for all layers | 64 |
| padding | [int,...?] | padding for each convolution layer, if single value is given it will be used for all layers. padding should be the (kernel-1)/2 on each layer | 5 |



## BIDIRECTIONAL LSTM PARAMS
| Parameter | type | Description | Default Value |
| --- | --- | --- | --- |
| hidden_size | int | number of hidden neurons for each LSTM layer | 200 |
| | | | |

## PRESNET PARAMS
| Parameter | type | Description | Default Value |
| --- | --- | --- | --- |
| resnet_type | string | specify type of Presnet from ['resnet18', 'resnet34', 'resnet50', 'resnet101', 'resnet152'] | resnet34 |
| kernel | [int,int] | kernel used for basisblocks and bottlenecks of PresNet | 11 |
| padding | [int,...?] | padding for basicblocks and bottlenecks. Padding should be the (kernel-1)/2 | 5 |



## EXAMPLE CONFIG FILE
```
#Example run config
silent=True
hide_ui=True
name=cnn_700_51k
evaluate_on_test=False
min_updates=20000
max_updates=20000
minibatch_size=32
eval_interval=100
learning_rate=0.0001
angle_lr=0.001
max_sequence_length=700
training_file=training_95
validation_file=validation
use_evolutionary=True
max_time=259200
loss_atoms=c_alpha

#Model parameters example for CNN
architecture=CNN
layers=5
output_size=60
#Specify kernel and padding for each layer or 1 value for same for all layers
kernel=51
padding=25
channels=200,500,500,200
dropout=0.5,0.5
#Dilation and stride can also be specified but will default to 1 for all layers


```


openprotein/config.py
# This file is part of the ProteinCNN project.
#
# @author Stig Killendahl & Kevin Jon Jensen
#
# Based on the OpenProtein framework, please see the LICENSE file in the root directory.

int_or_none = lambda x: int(x) if x is not None else None
str_to_bool = lambda x: x.lower() == "true"
int_list    = lambda x: [] if x == '' else [int(v.strip()) for v in x.split(",")]
float_list  = lambda x: [float(v.strip()) for v in x.split(",")]

class RunConfig():

    def __init__(self, file=None):
        config = {}
        if file is not None:
            config = self._create_map(file)
        self._create_config(config)

    def _create_map(self, filepath):
        configs = {}
        with open(filepath) as f:
            content = f.readlines()

        for line in content:

            if line[0] is "#" or line[0] == '\n':
                continue
            line = line.split("=")
            param, value = line[0], line[1]
            configs[param.strip()] = value.strip()
        return configs

    def _create_config(self, config):
        self.run_params = {
            "name":                 config.get("name", ""),
            "silent" :              str_to_bool(config.get("silent", True)),
            "hide_ui":              str_to_bool(config.get("hide_ui", False)),
            "evaluate_on_test":     str_to_bool(config.get("evaluate_on_test", False)),
            "min_updates":          int(config.get("min_updates", 1000)),
            "max_updates":          int(config.get("max_updates", 2000)),
            "minibatch_size":       int(config.get("minibatch_size", 1)),
            "eval_interval":        int(config.get("eval_interval", 10)),
            "learning_rate":        float(config.get("learning_rate", 0.01)),
            "max_sequence_length":  int(config.get("max_sequence_length", 2000)),
            "n_proteins":           int_or_none(config.get("n_proteins", None)),
            "use_evolutionary":     str_to_bool(config.get("use_evolutionary", True)),
            "training_file":        config.get("training_file","data/preprocessed/sample.txt"),
            "validation_file":      config.get("validation_file", "data/preprocessed/sample.txt"),
            "testing_file":         config.get("testing_file", "data/preprocessed/sample.txt"),
            "max_time":             int_or_none(config.get("max_time", None)),
            "c_epochs":             int(config.get("c_epochs", 0)),
            "only_angular_loss":    str_to_bool(config.get("only_angular_loss",False)),
            "angles_lr":            float(config.get("angles_lr", 0.01)),
            "loss_atoms":           config.get("loss_atoms", "c_alpha"),
            "weight_decay":         float(config.get("weight_decay", 0.0)),
            "betas":                float_list(config.get("betas","0.95,0.99" )),
            "print_model_summary":  str_to_bool(config.get("print_model_summary", "False")),
            "max_restarts":         int(config.get('max_restarts', 0)),
            "breakpoints":          int_list(config.get('breakpoints', ''))
        }

        self.model_params = {
            "architecture":         config.get("architecture", "cnn").lower(),
            "kernel":               int_list(config.get("kernel", "11")),
            "padding":              int_list(config.get("padding", "5")),
            "output_size":          int(config.get("output_size", 60)),
            "hidden_size":          int(config.get("hidden_size", 25)),
            "layers":               int(config.get("layers", 4)),
            "dilation":             int_list(config.get("dilation", "1")),
            "stride":               int_list(config.get("stride", "1")),
            "resnet_type":          config.get("resnet_type", "resnet34"),
            "channels":             int_list(config.get("channels","64,125,250")),
            "dropout":              float_list(config.get("dropout" ,"0.0,0.0"),),
            "spatial_dropout":      str_to_bool(config.get("spatial_dropout", "False"))
        }
